server/rest/src/main/java/org/infinispan/rest/resources/LoggingResource.java

```diff
@@ -60,23 +61,33 @@ public Invocations getInvocations() {
       if (level == null && appenders == null) {
          return completedFuture(new NettyRestResponse.Builder().status(HttpResponseStatus.BAD_REQUEST).build());
       }
-      return taskManager.runTask("@@logging@set",
-            new TaskContext()
-                  .addOptionalParameter("loggerName", loggerName)
-                  .addOptionalParameter("level", level)
-                  .addOptionalParameter("appenders", appenders)
-                  .subject(request.getSubject())
-      ).handle((o, t) -> handle(t));
+      NonRecursiveEventLoopGroup.reserveCurrentThread();
```

Comment:
why do you need this for tasks? they are not blocking and this thread does not wait for the result.

Reply:
It isn't just about blocking vs non blocking. The problem is the combination of netty event loop picking a specific thread to run (which can include one that was previously used) on during submission and the fact that not all of our code is non blocking (the `runTask` invokes blocking code).

So what can happen is in this case non blocking thread 2 runs this code which calls `engine.handles(String)` which does a remote invocation that it blocks on. If the remote execution then calls back into this node and selects thread 2 to handle the request it deadlocks, since thread 2 is blocking waiting it can't run a new task. The `NonRecursiveEventLoopGroup.reserveCurrentThread` makes it so that the current thread can't be chosen by the netty event loop until it is released. This is just a workaround until the entire method can be made non blocking. Due to the deadlock nature, it can't be solved purely by not having a block hound exception.

Reply:
`reserveCurrentThread` seems to "reserve" an `executor` and not a single thread. 
I guess if you invoke `reserveCurrentThread` recursively, eventually you are out of "executors".

Reply:
> `reserveCurrentThread` seems to "reserve" an `executor` and not a single thread.
With netty a given event loop is tied to a specific thread, so it is.

> I guess if you invoke `reserveCurrentThread` recursively, eventually you are out of "executors".
Yes, that is entirely possible to run out of event loops and thus threads if it was overused. That is the reason we want to use this very sparingly.